File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        answer[0] = 1;

        // forward iteration
        for (int i = 0; i < nums.length -1; i++) {
            answer[i+1] = answer[i] * nums[i];
        }

        // backward iteration
        int backwardMultiplier = nums[nums.length-1];
        for (int i = answer.length-2; i >= 0; i--) {
            answer[i] *= backwardMultiplier;
            backwardMultiplier *= nums[i];
        }

        return answer;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
	ans := make([]int, len(nums))
	ans[0] = 1

	// forward iteration
	for i := 0; i < len(nums)-1; i++ {
		ans[i+1] = ans[i] * nums[i]
	}

	// backward iteration
	backwardMultiplier := nums[len(nums)-1]
	for i := len(nums) - 2; i >= 0; i-- {
		ans[i] *= backwardMultiplier
		backwardMultiplier *= nums[i]
	}

	return ans
}